compiler/rustc_data_structures/src/thin_vec.rs

```diff
@@ -46,6 +56,53 @@ impl<T> ::std::ops::DerefMut for ThinVec<T> {
     }
 }
 
+impl<T> FromIterator<T> for ThinVec<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut iter = iter.into_iter();
+
+        match iter.next() {
+            None => ThinVec(None),
+            Some(first) => {
+                // + 1 for the first element
+                let mut vec = Vec::with_capacity(iter.size_hint().0 + 1);
+
+                vec.push(first);
+                vec.extend(iter);
+
+                ThinVec(Some(Box::new(vec)))
+            }
+        }
+    }
+}
+
+impl<T> IntoIterator for ThinVec<T> {
+    type Item = T;
+    type IntoIter = std::vec::IntoIter<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        // This is still performant because `Vec::new()` does not allocate.
+        self.0.map_or_else(Vec::new, |ptr| *ptr).into_iter()
```

Comment:
```suggestion
        self.0.as_deref().unwrap_or_default().into_iter()
```

Reply:
I don't think that will work because `as_deref` will return `&[T]`.

Reply:
It will return `&Vec`, but you're right that because it's a borrow it won't work.